oauth2.md

```diff
@@ -0,0 +1,205 @@
+| 302      | **Valid** `redirect_uri`. Errors are redirected to the provided `redirect_uri`. | (Refer to Sample 1 below) |
+| 400      | **Invalid** `redirect_uri`. Errors are directly returned as json body payload.  | (Refer to Sample 2 below) |
+
+**Note:** `redirect_uri` is only considered as valid, if the value used is registered under the corresponding `client_id` (Partner) used.
+
+```
+// Sample 1
+https://your-app.example.com/callback?error=unsupported_response_type&error_description=only+supports%3A+code&state=dK6fRBNVY1HM2GQ4aBwM
+```
+
+```
+// Sample 2
+{
+  "error": "invalid_request",
+  "error_description": "invalid client_id",
+  "state": "dK6fRBNVY1HM2GQ4aBwM"
+}
+```
+
+### 3. User Authorization
+
+The user will be redirected to log in and authorize your app on Mocaverse's site.
+
+### 4. Authorization Code Redirect Callback
+
+If the user approves the authorization request, they will be redirected back to your `redirect_uri` containing a one-time use authorization `code` and the provided `state` from step 2.
+
+```
+https://your-app.example.com/callback?code=602cbbe1-60dd-482f-b0d8-329f5f1254c7&state=v6glrJn3gf3qL4rPFLBB
+```
+
+If the user denies the authorization request, the user will be redirected to your `redirect_uri` containing an `access_denied` error and the provided `state` from step 2.
+
+```
+https://your-app.example.com/callback?error=access_denied&state=E0BXkRkKnvqMiDdYC8MW
+```
+
+Important: Validating the `state` if the value matches with the initial request during the User Authorization Request is strongly recommended.
+
+### 5. Access Token Request
+
+#### Request
+
+Request the `access_token` via `POST https://oauth.mocaverse.xyz/api/oauth/token` with the following `application/x-www-form-urlencoded` parameters:
+
+- `grant_type`: must be `authorization_code`
+- `client_id`: must be the provided `client_id` from step 1
+- `client_secret`: must be the provided `client_secret` from step 1
+- `code`: the authorization code received from step 4
+- `redirect_uri` must match the same value used during **Authorization Request**
+
+#### Response
+
+| Response | Description | Sample Return             |
+|----------|-------------|---------------------------|
+| 200      | Successful  | (Refer to Sample 1 below) |
+| 400      | Bad Request | (Refer to Sample 2 below) |
+
+```
+// Sample 1
+{
+  "access_token": "...",
+  "token_type": "bearer"
+}
+```
+
+```
+// Sample 2
+{
+  "error": "invalid_request",
+  "error_description": "invalid auth code"
+}
+```
+
+## OAuth2 APIs
+
+### Partner User Registration
+
+Link a user id on your platform to a user in Mocaverse.
+
+#### Request
+
+```
+POST https://api.moca-id.mocaverse.xyz/api/partner/add-user
+```
+
+Request Headers:
+
+- `Authorization: Bearer <access_token>`
+- `Content-Type: application/json`
+
+Request Body Parameters:
+
+- `partnerUserId`: ID of the user on your platform
+- `partnerUserName` (*Optional*): Username of the user on your platform
```

Comment:
use the same description as partner widget:
partnerUserId: Their unqiue User ID in your system (e.g. 2c55dc7d-5715-4cb9-9801-f62498c02077)
partnerUserName: Their Display name in your system (e.g. gamertag123)